doc/compiler/optimizer/OptimizerDesignFeatures.md

```diff
@@ -0,0 +1,303 @@
+Exception [1] and GNU General Public License, version 2 with the
+OpenJDK Assembly Exception [2].
+
+[1] https://www.gnu.org/software/classpath/license.html
+[2] http://openjdk.java.net/legal/assembly-exception.html
+
+SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+-->
+
+# Design Features in the OpenJ9 JIT Optimizer
+
+The design of an optimizer, specifically for a JIT compiler, must consider many aspects that differentiate
+it from a static compiler.
+
+A static compiler has the required compile time budget to run sophisticated analyses and often repeats them
+many times to achieve excellent code quality, whereas a JIT compiler must balance compile time and run time
+performance.
+
+The other major difference is that the JIT compiler runs concurrently with the application and can therefore
+keep track of the global state of the program relevant for optimization, such as the class hierarchy and
+profiling data.
+
+Most runtimes that have a JIT compiler typically also have an interpreter, which is beneficial from a
+performance standpoint (infrequently executed code needs not incur compilation overhead), as well as from a
+functional standpoint (edge cases that are not worth handling in the JIT compiler can be delegated to the
+interpreter).
+
+Allowing a "mixed mode" execution model, i.e., compiled code and interpreter running different parts of the
+application, provides flexibility in terms of selecting which methods ought to be compiled and when to
+compile them.
+
+Given these differences between static compilation and JIT compilation, different aspects of compiler design
+and infrastructure gain in importance depending on the context.
+
+Features that we feel are especially important from an optimization perspective are
+
+- [Heuristics on what/when/how to compile](#heuristics)
+- [Hot path information and profile guided optimization in general](#hot-path-information-and-profile-guided-optimization)
+- [Speculative optimization and compensation techniques as fallback mechanisms when speculation fails](#speculative-optimization-and-compensation-techniques-as-fallback-mechanisms)
+- [Avoidance of expensive analyses such as Static Single Assignment(SSA), interprocedural analysis, expensive aliasing analysis, etc.](#avoidance-of-expensive-analyses)
+- [IL representation that is more amenable to local analyses](#il-representation)
+- [Cooperative threading model, compiled method meta data in conjunction with a stack walker](#cooperative-threading-model-and-compiled-method-meta-data)
+
+## Heuristics
+
+Methods are the units of compilation in the OpenJ9 JIT compiler.
+
+In the OpenJ9 JIT compiler, compilations are triggered either based on a method having been [invoked a certain number of times](https://github.com/eclipse-openj9/openj9/blob/20c2e83228780955fd62e1620160a81e7cbec41e/runtime/compiler/control/CompilationRuntime.hpp#L525)
```

Comment:
Some of the information in this section looks like it overlaps with information in the [Compilation Control](https://github.com/eclipse-openj9/openj9/blob/master/doc/compiler/control/CompilationControl.md) documentation.  I wonder if it should be consolidated with that section to avoid redundancy, and to make it less likely that one will get out-of-date if the other changes.

Reply:
Just to be clear, by "this section", I meant the Heuristics section.

Reply:
I think I should link the key concepts mentioned in this Heuristics section to [CompilationControl.md](https://github.com/eclipse-openj9/openj9/blob/master/doc/compiler/control/CompilationControl.md) such as invocation counts, compilation threads, Sampler Thread, etc.